### ISSUE TITLE
Infinite recursion / PHP memory exhaustion in SetDefaultSelectedPaymentMethod when default method is set and quote has trigger_recollect

### ISSUE THREAD

**Issue by rjones-indez** [mollie/magento2-hyva-checkout#93]
Hi Guys, We were having an issue with a client and an infinite loop bug relating to one of the observers in the module, where the default payment method was set. I've asked Claude to outline the issue below. We've got a work around via another observer as a temporary measure, but it might be something you want to look at.

**Package:** `mollie/magento2-hyva-checkout` — reproduced on 1.9.1, and confirmed still present in **3.0.1** by code inspection.
**Gateway:** `mollie/magento2` 2.49.0 (also checked 3.1.2).
**Magento:** 2.4.6-p15 CE, Hyvä. **PHP:** 8.1.

#### Summary

`Observer/SalesQuoteCollectTotalsBefore/SetDefaultSelectedPaymentMethod` calls
`PaymentMethodManagement::set()` during the `sales_quote_collect_totals_before` event. That call reloads and saves the quote, which runs `collectTotals()` again, which re-fires the same event, which runs the observer again. When the quote is flagged `trigger_recollect = 1`, this becomes an
**unbounded quote save/reload cycle that exhausts PHP memory** (2 GB+) instead of returning cleanly.

This is **related to #69 but a distinct variant**. The #69 fix (`isMethodActive()` check, released in 1.8.3) does not cover this case, because here the configured default method *is* active — so execution passes that guard and proceeds to `set()`, then recurses.

#### Still present in the latest release

I audited the full **3.0.1** package (both `di.xml`, `events.xml`, every PHP file) and the paired gateway **3.1.2**. The loop path is unchanged from 1.9.1 — the only difference in the observer is an added `isModuleEnabled()` early-return, which is irrelevant when the module is enabled. Specifically, in 3.0.1:

- the observer is still registered on `sales_quote_collect_totals_before`;
- it still calls `PaymentMethodManagement::set($quote->getId(), $payment)` during totals collection;
- it still catches and **silently swallows** `InvalidTransitionException`;
- there is still **no re-entrancy guard**;
- nothing else in the package touches the totals-collection / `PaymentMethodManagement` path.

So upgrading to 3.0.1 reproduces the OOM identically.

#### Trigger conditions (all true together)

1. `payment/mollie_general/default_selected_method` is set to a real, **active** method (e.g. `mollie_methods_creditcard`).
2. Another module restricts the available payment methods (in our case a plugin on `Magento\Payment\Model\PaymentMethodList::getList`). Same precondition noted in #69.
3. The quote has `trigger_recollect = 1`. This is not exotic — Magento sets it automatically after a **catalog price-rule reindex** and on **currency/store switch**, so it hits real carts in production.
4. The cart has no valid address yet (so `set()` throws `InvalidTransitionException`, which the observer swallows).

#### Apparent mechanism

- `trigger_recollect = 1` makes `Quote::_afterLoad()` run `collectTotals()` during a quote load.
- `collectTotals()` dispatches `sales_quote_collect_totals_before` → this observer.
- The default method is active, so it passes the 1.8.3 guard and calls `PaymentMethodManagement::set()`.
- `set()` reloads + saves the quote → `collectTotals()` runs again → the event re-fires → observer again.
- On an address-less cart `set()` throws `InvalidTransitionException`; the observer swallows it, so the payment method is never persisted, `quoteHasActivePaymentMethod()` never becomes true, and nothing terminates the loop.
- Result: recursive quote saves until the PHP memory limit is hit.

#### Error signature

```
Fatal error: Allowed memory size of 2147483648 bytes exhausted ...
  in .../Model/ResourceModel/Db/VersionControl/Snapshot.php
  / .../Model/AbstractExtensibleModel.php
  / colinmollenhour/php-redis-session-abstract/.../Handler.php

PHP Warning: Unknown: Cannot call session save handler in a recursive manner
```

Observer frame:

```
.../Observer/SalesQuoteCollectTotalsBefore/SetDefaultSelectedPaymentMethod.php → Magento\Quote\Model\PaymentMethodManagement->set()
```

#### Steps to reproduce

1. Set a default selected method to an active method (e.g. `mollie_methods_creditcard`).
2. Have a module active that filters `PaymentMethodList::getList`.
3. Add an item to a guest cart; do **not** select a payment method or address.
4. Flag the quote: `UPDATE quote SET trigger_recollect = 1 WHERE entity_id = <id>;`
   (or trigger a catalog price-rule reindex / currency switch).
5. Load the cart/checkout page → OOM.

Setting `default_selected_method` to empty avoids it entirely (the observer early-returns), which confirms the observer is the source.

#### Suggested fix direction

The observer performs a payment-method **write** (`PaymentMethodManagement::set()`, which saves the quote) inside `sales_quote_collect_totals_before`. Either:

- guard against re-entrancy so a `set()`-triggered recollect can't re-run the observer, and/or
- don't attempt to persist the default method when the quote can't accept it yet (the `InvalidTransitionException` path) — swallowing the failure leaves the loop with no terminator.

#### Workaround (for reference)

We disable this observer via `events.xml` and replace it with a thin wrapper that delegates to it behind a static re-entrancy flag: the outer call runs the real logic (default method is still applied), and any nested re-fire it triggers is skipped. Peak memory drops from 2 GB+ (OOM) to ~265 MB and the loop no longer occurs.

**Comment by Frank-Magmodules** [comment]
Hi there @rjones-indez,

I noticed this issue was still open, but since it has been covered in the latest release ([v3.0.2](https://github.com/mollie/magento2-hyva-checkout/releases/tag/v3.0.2)), I’m closing it for now.
Feel free to reopen it if you’re still experiencing any issues or think further action is needed.

Have a great start to the week!
